tests/run-with-compiler-custom-args/staged-streams_1.scala

```diff
@@ -0,0 +1,697 @@
+      * @tparam A        the type of the producer's elements.
+      * @return          the enhanced producer
+      */
+    private def addCounter[A](n: Expr[Int], producer: Producer[A]): Producer[(Var[Int], A)] = {
+      new Producer[(Var[Int], A)] {
+        type St = (Var[Int], producer.St)
+        val card = producer.card
+
+        def init(k: St => Expr[Unit]): Expr[Unit] = {
+          producer.init(st => {
+            Var(n) { counter =>
+              k(counter, st)
+            }
+          })
+        }
+
+        def step(st: St, k: (((Var[Int], A)) => Expr[Unit])): Expr[Unit] = {
+          val (counter, currentState) = st
+          producer.step(currentState, el => '{
+            ~k((counter, el))
+          })
+        }
+
+        def hasNext(st: St): Expr[Boolean] = {
+          val (counter, currentState) = st
+          producer.card match {
+            case Many => '{ ~counter.get > 0 && ~producer.hasNext(currentState) }
+            case AtMost1 => '{ ~producer.hasNext(currentState) }
+          }
+        }
+      }
+    }
+
+    /** The nested stream receives the same variable reference; thus all streams decrement the same global count.
+      *
+      * @param  n      code of the variable to be threaded to the downstream.
+      * @param  stream the upstream to enhance.
+      * @tparam A      the type of the producer's elements.
+      * @return        a linear or nested stream aware of the variable reference to decrement.
+      */
+    private def takeRaw[A: Type](n: Expr[Int], stream: StagedStream[A]): StagedStream[A] = {
+      stream match {
+        case linear: Linear[A] => {
+          val enhancedProducer: Producer[(Var[Int], A)] = addCounter[A](n, linear.producer)
+          val enhancedStream: Linear[(Var[Int], A)] = Linear(enhancedProducer)
+
+          // Map an enhanced stream to a stream that produces the elements. Before
+          // invoking the continuation for the element, "use" the variable accordingly.
+          mapRaw[(Var[Int], A), A]((t: (Var[Int], A)) => k => '{
+            ~t._1.update('{~t._1.get - 1})
+            ~k(t._2)
+          }, enhancedStream)
+        }
+        case nested: Nested[A, bt] => {
+          val enhancedProducer: Producer[(Var[Int], bt)] = addCounter[bt](n, nested.producer)
+
+          Nested(enhancedProducer, (t: (Var[Int], bt)) => {
+            // Before invoking the continuation for the element, "use" the variable accordingly.
+            // In contrast to the linear case, the variable is initialized in the originating stream.
+            mapRaw[A, A]((el => k => '{
+              ~t._1.update('{~t._1.get - 1})
+              ~k(el)
+            }), addTerminationCondition(b => '{ ~t._1.get > 0 && ~b}, nested.nestedf(t._2)))
+          })
+        }
+      }
+     }
+
+    /** A stream containing the first `n` elements of this stream. */
+    def take(n: Expr[Int]): Stream[A] = Stream(takeRaw[Expr[A]](n, stream))
```

Comment:
```scala
def take(n: Int): Stream[A] = Stream(takeRaw[Expr[A]](n.toExpr, stream))
```

Reply:
@nicolasstucki what’s the advantage of your proposal?

Reply:
The aim is to hide the `Expr` from the end user.

Reply:
Then do we want `fold` to have the following signature?

~~~ scala
def fold[W : Type](z: W, f: (W, A) => W): W
~~~

I’m not sure this is what we want?

Reply:
You are right, it is not exactly what I want. We need to hide the Expr behind a macro. I will test my signature idea and then document it here if it works